pkg/vm/vm.go

```diff
@@ -693,27 +696,48 @@ func (v *VM) execute(ctx *Context, op Instruction, parameter []byte) (err error)
 	// Numeric operations.
 	case ADD:
 		a := v.estack.Pop().BigInt()
+		v.checkBigIntSize(a)
```

Comment:
Maybe we can hide in the `BigInt()` itself?

Reply:
We definitely can, but I think that later all limits will become VM parameters, so an item conversion function will have no sufficient context to perform this check.

Reply:
It can be refactored if we decide to do that (passing limits around for example), but at the moment you're copy-pasting a lot of these checks and it's hard to ensure that we're not missing anything. And at the same time even now `checkBigIntSize` has no real connection to VM, it easily can be a function and not a method. So it can be invoked from `makeStackItem` and `BigInt` which is less code and more reliability at the same time.

And it's not that if we're to decide to make limits configurable at the VM level that we could do it right now easily, `MaxItemSize` and `MaxArraySize` are not bound by the `VM` at the moment, they're package-level, so doing the same thing with `MaxBigIntegerSizeBits` doesn't make anything worse, I think.

Reply:
I agree that what you suggest can be a cleaner implementation.
However I am thinking not about difficulty of passing limits around, but rather about strict conformance to _the_ reference implementation. For example, there are no limits checking in `MAX` opcode, so if we make it as you said, this script will behave differently in our (fail, probably correctly) and C# NEO (succeed and return 0) implementations.
```
PUSH1
PUSHBYTES32
CAT         // create ByteArray which converts to 2^32
PUSH0
MIN         // min(2^32, 0) == 0, but we panic here.
```
I suggest to leave implementation as it is now and create an issue to refactor this and fix all bugs of reference implementation in NEO3.0 . What do you think? @roman-khimov @volekerb

Reply:
> For example, there are no limits checking in MAX opcode

Oh, my... That's exactly the problem I was talking about, it's so easy to miss something with this kind of implementation.

> I suggest to leave implementation as it is now and create an issue to refactor this and fix all bugs of reference implementation in NEO3.0

OK, let's do that, at the moment NEO 2.0 compatibility probably is more important for us, but we need a bug and we probably will need to file a bug in the neo-vm project also.